### PR TITLE
Automatically determine interface from constructed quantum script

### PR DESCRIPTION
**Pull request by AlbertMitjans**
The `QNode` class now accepts an ``auto`` interface, which automatically detects the interface of the given input.
```python
dev = qml.device("default.qubit", wires=2)

@qml.qnode(dev, interface="auto")
def circuit(weight):
    qml.RX(weight[0], wires=0)
    qml.RY(weight[1], wires=1)
    return qml.expval(qml.PauliZ(0))

interface_tensors = [
    np.array([0, 1]),
    torch.Tensor([0, 1]),
    tf.Variable([0, 1], dtype=float),
    jnp.array([0, 1]),
    [0, 1],
]

for tensor in interface_tensors:
    res = circuit(weight=tensor)
    print(res, type(res))
```
Prints
```
1.0 <class 'pennylane.numpy.tensor.tensor'>
tensor(1.0000, dtype=torch.float64) <class 'torch.Tensor'>
tf.Tensor(0.9999999896704281, shape=(), dtype=float64) <class 'tensorflow.python.framework.ops.EagerTensor'>
1.0 <class 'jaxlib.xla_extension.DeviceArray'>
1.0 <class 'pennylane.numpy.tensor.tensor'>
```

### CODE REVIEW

**Comment by AlbertMitjans** [comment]
[sc-26970]

File: pennylane/interfaces/execution.py
Comment (antalszava) [inline pennylane/interfaces/execution.py:24]: Why was this required?

File: pennylane/interfaces/execution.py
Comment (antalszava) [inline pennylane/interfaces/execution.py:42]: What happens if `qml.execute` is called directly without using a QNode? Good to add tests for that case too.

Also, should `"auto"` be our default going forward?

**Review by antalszava — COMMENTED** [review]
`"auto"` is introduced for `qml.execute`, but the logic of `qml.execute` was not updated to support it. Why?

`qml.execute` may be called at other parts of the codebase, would those cases be supported?

File: pennylane/interfaces/execution.py
Comment (AlbertMitjans) [reply pennylane/interfaces/execution.py:42]: Added "auto" interface for all interface checks.

Yes, `"auto"` will be our default going forward.

File: doc/releases/changelog-dev.md
Comment (antalszava) [inline doc/releases/changelog-dev.md:69]: The suggestion is just to group the "vanilla" cases first.

The output here reads a bit busy, maybe good to add a sentence right before the example explaining that we can use the same QNode with gate parameters from different frameworks and that this is being showcased with multiple input parameters.
```suggestion
  >>> interface_tensors = [[0, 1], np.array([0, 1]), torch.Tensor([0, 1]), tf.Variable([0, 1], dtype=float), jnp.array([0, 1])]
  >>> for tensor in interface_tensors:
        res = circuit(weight=tensor)
        print(res, type(res))
  1.0 <class 'pennylane.numpy.tensor.tensor'>
  1.0 <class 'pennylane.numpy.tensor.tensor'>
  tensor(1.0000, dtype=torch.float64) <class 'torch.Tensor'>
  tf.Tensor(0.9999999896704281, shape=(), dtype=float64) <class 'tensorflow.python.framework.ops.EagerTensor'>
  1.0 <class 'jaxlib.xla_extension.DeviceArray'>
```

File: doc/releases/changelog-dev.md
Comment (antalszava) [inline doc/releases/changelog-dev.md:60]: The formatting is a bit off here, lines of code after lines starting with `>>>` should start with `...`

image

An alternative is to first have a python block and then a pycon block.

File: pennylane/qnode.py
Comment (antalszava) [inline pennylane/qnode.py:225]: Why was this required? It seems like we're not setting the interface to `None` `if self.diff_method is None` by omitting this call.

File: pennylane/qnode.py
Comment (antalszava) [inline pennylane/qnode.py:726]: Should this be moved further up to line 714 or so? Seems like this is not being executed in all cases because the method may have returned a value before already.

File: tests/interfaces/test_tensorflow.py
Comment (antalszava) [inline tests/interfaces/test_tensorflow.py:697]: The suggestion on rewording applies to other occurrences too.
```suggestion
            pytest.skip("Can't detect interface without a parametrized gate in the tape")
```

**Review by antalszava — CHANGES_REQUESTED** [review]
Hi @AlbertMitjans, this is getting there! :slightly_smiling_face: :runner:  A couple of more comments, there may be some more edge cases to explore (one more pressing is to determine where in the QNode resetting the `interface` attribute happens).

There may be many edge cases that would be covered by running the test suite with `interface="auto"` by default. This could motivate our plan with making `"auto"` the default.

File: tests/interfaces/test_tensorflow.py
Comment (AlbertMitjans) [reply tests/interfaces/test_tensorflow.py]: Fixed!

File: pennylane/qnode.py
Comment (AlbertMitjans) [reply pennylane/qnode.py]: Good catch! I moved this code right after computing the results, and also added it inside the new returns type branch.

File: pennylane/qnode.py
Comment (AlbertMitjans) [reply pennylane/qnode.py]: Because we first need to determine the interface before calling `self._update_gradient_fn()` with "auto".

But you are right! I moved the `self.interface != "auto"` check inside the `self._update_gradient_fn` method.

File: doc/releases/changelog-dev.md
Comment (AlbertMitjans) [reply doc/releases/changelog-dev.md:60]: Added python block

File: doc/releases/changelog-dev.md
Comment (AlbertMitjans) [reply doc/releases/changelog-dev.md]: Grouped vanilla cases.

Add sentence and use prettier output.

File: pennylane/interfaces/execution.py
Comment (AlbertMitjans) [reply pennylane/interfaces/execution.py:574]: It is "scipy"

I added `"scipy": "autograd"` to the interface mapping in the `execution.py` file to tackle this corner case.

Comment (antalszava) [reply pennylane/interfaces/execution.py:574]: What if the circuit has no parametrized gates, but a SparseHamiltonian taking a sparse matrix?

> If needed we can add "scipy": "numpy" or "scipy": "autograd" in the interface mapping of the execution.py file

I think that would be safest or handling the `scipy` case in some way would be great. :+1:

File: doc/releases/changelog-dev.md
Comment (antalszava) [inline doc/releases/changelog-dev.md:68]:
```suggestion
  for tensor in interface_tensors:
      res = circuit(weight=tensor)
      print(f"Result value: {res:.2f}; Result type: {type(res)}")
```

File: doc/releases/changelog-dev.md
Comment (antalszava) [inline doc/releases/changelog-dev.md:57]: The example looks great! :100:

**Review by antalszava — COMMENTED** [review]
:clap: :clap: :clap: 

Ironing out the SciPy option is the last thing, otherwise, this is :1st_place_medal: 

Maybe it would be good to have another reviewer - this is an important change and best not to miss anything.

File: pennylane/qnode.py
Comment (AlbertMitjans) [reply pennylane/qnode.py:638]: Added tests with the new return types.

File: pennylane/interfaces/execution.py
Comment (AlbertMitjans) [reply pennylane/interfaces/execution.py:575]: Done! Added "scipy" -> "autograd" interface mapping + added test

**Review by antalszava — APPROVED** [review]
Looks good! :muscle: :clap: This is amazing, will be very useful. :heart_eyes: 

A condition on this approval is to have another approval, otherwise, things look great. :slightly_smiling_face:

**Comment by AlbertMitjans** [comment]
@Jaybsoni I will wait for your approval before merging

**Comment by rmoyard** [comment]
Hi @AlbertMitjans, great changes! One thing that should be done is adding those changes to `_execute_new` in the `execution.py` file. Also do you think we could change the default interface to `auto`?

**Comment by AlbertMitjans** [comment]
> One thing that should be done is adding those changes to _execute_new in the execution.py file.
We are calling `_execute_new` with the updated interface, thus I believe we don't need to add any logic there.

> Also do you think we could change the default interface to auto?
@albi3ro mentioned that we should first try this out as an optional feature to make sure it works correctly. If it does we can make it the default.

**Comment by rmoyard** [comment]
@AlbertMitjans Thanks for the second point! Our methodology for the return type project is to duplicate the function (one public and one private). Right at the beginning, the public one calls the private one. The private one should be able to replace the public one without having to double check if something is missing (when we will deprecate it). So in this case I would suggest to move the code bloc after the `qml.active_return` and add the bloc in the function `_execute_new`.

**Review by antalszava — CHANGES_REQUESTED** [review]
As per Romain's find, will just double-check that the new return-type pipeline works as expected.

**Comment by rmoyard** [comment]
Thanks @AlbertMitjans ! 💯  You can add `["autograd", "auto"]` in the test file `test_qnode_autograd_new.py` and `test_autograd_new.py`, it should solve the coverage issues.

**Comment by AlbertMitjans** [comment]
> You can add ["autograd", "auto"] in the test file test_qnode_autograd_new.py and test_autograd_new.py, it should solve the coverage issues.

Added!

**Review by albi3ro — COMMENTED** [review]
The docstring for `QNode` and `execute` should be updated to include the option `"auto"`.

**Comment by AlbertMitjans** [comment]
> The docstring for `QNode` and `execute` should be updated to include the option `"auto"`.

Done!

File: pennylane/interfaces/execution.py
Comment (albi3ro) [inline pennylane/interfaces/execution.py:45]:
```suggestion
    "scipy": "Numpy",
```
Scipy wouldn't be trainable, so it should map to the non-trainable option.

File: pennylane/qnode.py
Comment (albi3ro) [inline pennylane/qnode.py:69]:
```suggestion
            * ``"auto"``: The QNode automatically detects the interface from the input values of
```

**Review by albi3ro — APPROVED** [review]
Glad to have this! 👍 Looks good to me.

**Review by antalszava — APPROVED** [review]
Looks good! :tada: Awesome feature, great work :slightly_smiling_face:  :heart_eyes:

**Pull request merged**
